test/partition-healing-tests.js

```diff
@@ -0,0 +1,377 @@
+                dsl.callEndpoint(t, tc, "/admin/healpartition/disco", {},
+                        function(eventBody) { validateHealRequest(t, tc, eventBody) }),
+                removeHealPartitionDiscoResponse(t, tc),
+
+                // SUT sends /join to b.
+                dsl.waitForJoins(t, tc, 1),
+
+                // Assert ping was sent to b with a1 and sut alive.
+                checkSutMergedWithB(t, tc),
+
+                // Make node membership global again, since partition is over.
+                function(list, cb) {
+                    delete tc.test_state['a1']['membership'];
+                    delete tc.test_state['b1']['membership'];
+                    delete tc.test_state['b2']['membership'];
+                    cb(list);
+                },
+                dsl.assertStats(t, tc, 4, 0, 0),
+            ]
+        })
+     );
+
+
+function waitForPingResponseFromTarget(t, tc) {
+    return function waitForPingResponseFromTarget(list, cb) {
+        return dsl.waitForPingResponse(t, tc, tc.test_state['target_idx'])(list, cb);
+    }
+}
+
+// Mark b1 and b2 faulty in a1.
+function markBFaultyInA(t, tc) {
+    return function markBFaultyInA(list, cb) {
+        var b1_port = tc.test_state['b1'].port,
+        b2_port = tc.test_state['b2'].port;
+        _.find(tc.test_state['a1'].membership, {port: b1_port}).status = 'faulty';
+        _.find(tc.test_state['a1'].membership, {port: b2_port}).status = 'faulty';
+        cb(list);
+    }
+}
+
+// Mark sut and a1 faulty in memberships of B.
+// b1.membership and b2.membership are referring to the same list.
+function markSutA1FaultyInB(t, tc) {
+    return function markSutA1FaultyInB(list, cb) {
+        var a1_port  = tc.test_state['a1'].port,
+        sut_port = tc.test_state['sut'].port;
+        _.find(tc.test_state['b1'].membership, {port: a1_port}).status = 'faulty';
+        _.find(tc.test_state['b1'].membership, {port: sut_port}).status = 'faulty';
+        cb(list);
+    }
+}
+
+// Increase incarnation number of B, both in node and global test state.
+function increaseInternalIncB(t, tc) {
+    return function increaseInternalIncB(list, cb) {
+        tc.test_state['b1'].incarnationNumber += 3;
+        tc.test_state['b2'].incarnationNumber += 3;
+
+        // keep the incarnation number in membership in sync with internal state
+        var b1_port = tc.test_state['b1'].port,
+            b2_port = tc.test_state['b2'].port;
+        _.find(tc.test_state['b1'].membership, {port: b1_port}).incarnationNumber += 3;
+        _.find(tc.test_state['b1'].membership, {port: b2_port}).incarnationNumber += 3;
+        cb(list);
+    }
+}
+
+function decreaseIncOfSutInB(t, tc) {
+    return function decreaseIncOfSutInB(list, cb) {
+        var B = tc.test_state['b1'].membership;
+        _.find(B, {port: tc.test_state['sut'].port}).incarnationNumber = 1000;
+        cb(list);
+    }
+}
+
+
+function splitMemberships(t, tc) {
```

Comment:
you are right, the previous implementation was terrible. I updated it to have no shared items any more. Thanks!